explainer.md

```diff
@@ -293,6 +293,65 @@ input.readable.pipeInto(demuxer.writable);
 muxer.readable.pipeInto(output.writable);
 ```
 
+## Detailed design discussion
+
+### Codec configuration
+
+Many codecs and encoder/decoder implementations are highly configurable. WebCodecs intends to support most of the configuration options available in codecs today to efficiently allow for advanced use cases.
+
+Configuration options are classified into two types:
+- **Parameters** are metadata required to construct a compliant bitstream. These are required when constructing the encoder/decoder and cannot be changed. For example, the VP9 profile.
+- **Settings** are configuration options that influence the behavior of the encoder but do not change the type of bitstream produced. For example, target bitrate.
+
+Settings are further classified into three types:
+- **Static codec settings** must be specified when constructing the encoder and cannot be changed without reinitializing the encoder.
+- **Dynamic codec settings** apply to the lifetime of the encoder and can be changed at any point. Dynamic settings must be lightweight and not require reinitializing the encoder (concretely, does not require a new key frame to be produced).
+- **Frame settings** apply only to specific input frames.
+
+WebCodecs will maintain a standard definition of parameters for each supported codec. Additionally, the specification will establish common encoder settings that apply across codecs and implementation. However, we expect many settings will be implementation-specific. These will be available behind a feature detection and configuration API (TODO: sketch this).
+
+#### Configuration examples
+
+Both encoder and decoder constructors take in the codec name and required parameters. Encoders additionally take in a dictionary of codec settings.
+
+```javascript
+const encoder = new VideoEncoder({
+  codec: 'VP9',
+  profile: '1',
+  settings: {
+    targetBitRate: 80_000,
+  },
+});
+```
+
+Codec settings can be changed on-the-fly by bundling the changed settings with the next input image. The changed settings will be applied before encoding the image and apply to subsequent images.
+
+```javascript
+const encoder = new VideoEncoder(...);
+const writer = encoder.writable.getWriter();
+writer.write({
+  imageData: ...,
+  timestamp: ...,
+  changeCodecSettings: {
+    targetBitRate: 50_000,
+  },
+});
+```
+
+Frame settings are also bundled with the next input image. These settings do not persist beyond encoding for the image on which they appear.
+
+```javascript
+const encoder = new VideoEncoder(...);
+const writer = encoder.writable.getWriter();
+writer.write({
+  imageData: ...,
+  timestamp: ...,
+  frameSettings: {
+    forceKeyFrame: true,
```

Comment:
Not all frames can be key frames, so should this be a "make the next available frame a key frame" dynamic codec setting?  Or should we expect the user of the API to always be smart enough to time it right?

Reply:
I was already having difficulty identifying frame settings, so I'm just going to remove the category. We can add it back later if it turns out to have use.